storage-proofs/src/merkle/tree.rs

```diff
@@ -0,0 +1,354 @@
+#![allow(clippy::len_without_is_empty)]
+
+use std::marker::PhantomData;
+use std::path::PathBuf;
+
+use anyhow::Result;
+use generic_array::typenum::{self, U0};
+use merkletree::hash::Hashable;
+use merkletree::merkle;
+use merkletree::merkle::FromIndexedParallelIterator;
+use merkletree::store::StoreConfig;
+use rayon::prelude::*;
+
+use crate::hasher::{Hasher, PoseidonArity};
+
+use super::*;
+
+/// Trait used to abstract over the way Merkle Trees are constructed and stored.
+pub trait MerkleTreeTrait: Send + Sync + std::fmt::Debug {
+    type Arity: 'static + PoseidonArity;
+    type SubTreeArity: 'static + PoseidonArity;
+    type TopTreeArity: 'static + PoseidonArity;
+    type Hasher: 'static + Hasher;
+    type Store: Store<<Self::Hasher as Hasher>::Domain>;
+    type Proof: MerkleProofTrait<
+        Hasher = Self::Hasher,
+        Arity = Self::Arity,
+        SubTreeArity = Self::SubTreeArity,
+        TopTreeArity = Self::TopTreeArity,
+    >;
+
+    /// Print a unique name for this configuration.
+    fn display() -> String;
+    /// Returns the root hash of the tree.
+    fn root(&self) -> <Self::Hasher as Hasher>::Domain;
+    /// Creates a merkle proof of the node at the given index.
+    fn gen_proof(&self, index: usize) -> Result<Self::Proof>;
+    fn gen_cached_proof(&self, i: usize, levels: usize) -> Result<Self::Proof>;
+    fn height(&self) -> usize;
+    fn leaves(&self) -> usize;
+    fn from_merkle(
+        tree: merkle::MerkleTree<
+            <Self::Hasher as Hasher>::Domain,
+            <Self::Hasher as Hasher>::Function,
+            Self::Store,
+            Self::Arity,
+            Self::SubTreeArity,
+            Self::TopTreeArity,
+        >,
+    ) -> Self;
+}
+
+pub struct MerkleTreeWrapper<
+    H: Hasher,
+    S: Store<<H as Hasher>::Domain>,
+    U: PoseidonArity,
+    V: PoseidonArity = typenum::U0,
+    W: PoseidonArity = typenum::U0,
+> {
+    pub inner: merkle::MerkleTree<<H as Hasher>::Domain, <H as Hasher>::Function, S, U, V, W>,
+    pub h: PhantomData<H>,
+}
+
+impl<
+        H: 'static + Hasher,
+        S: Store<<H as Hasher>::Domain>,
+        U: 'static + PoseidonArity,
+        V: 'static + PoseidonArity,
+        W: 'static + PoseidonArity,
+    > MerkleTreeTrait for MerkleTreeWrapper<H, S, U, V, W>
```

Comment:
Consider more descriptive names when cleaning this up later. Not blocking for release.